[ML] Extend unit testing and make some improvements to deal with changes in seasonal components

tveasey opened the pull request:
In the course of other testing related to #6, I found evidence of potential instability in the decomposition when the seasonal components in the time series abruptly change. In addition, our lifecycle management of seasonal components was not reliably removing components which were no longer improving predictions. This was related to the way we were trying to preserve components which carry information about periodic variance in the data.

This adds a unit test to exercise these problems and makes some related improvements. As part of these I've implemented a failsafe to remove all seasonal components and start again if instability occurs. The current behaviour on this unit test (red actuals, blue predictions) is as follows:
![screen shot 2018-05-10 at 18 11 58](https://user-images.githubusercontent.com/7591487/39916718-3b49a4e4-5503-11e8-89b8-11119d625bdc.png)
![screen shot 2018-05-10 at 18 14 53](https://user-images.githubusercontent.com/7591487/39916722-3ff751a8-5503-11e8-9af4-55ac0a42fc3d.png)
![screen shot 2018-05-10 at 18 15 42](https://user-images.githubusercontent.com/7591487/39916727-4577cf40-5503-11e8-9442-5d44b8f61b48.png)
![screen shot 2018-05-10 at 18 17 16](https://user-images.githubusercontent.com/7591487/39916745-573025a2-5503-11e8-8fa4-f2c9cb069ea1.png)
![screen shot 2018-05-10 at 18 17 56](https://user-images.githubusercontent.com/7591487/39916750-59f15266-5503-11e8-99ca-1be7806804b3.png)

For reference, this is what was happening before:
![screen shot 2018-05-15 at 09 56 29](https://user-images.githubusercontent.com/7591487/40047080-6835ede8-5826-11e8-9bfd-08a1f4df500e.png)

This can affect results on metric and count analysis with seasonal signals where seasonality changes significantly from time to time.

droberts195 commented on lib/maths/CTimeSeriesDecompositionDetail.cc:
```diff
@@ -255,6 +258,7 @@ const TSizeVecVec SC_TRANSITION_FUNCTION{
     TSizeVec{SC_NORMAL, SC_NORMAL, SC_NORMAL, SC_NORMAL}};
 
 const std::string VERSION_6_3_TAG("6.3");
+const std::string VERSION_6_4_TAG("6.4");
```
We probably want to prevent a job that has written 6.4 state from running on a 6.3 node in a mixed 6.3/6.4 cluster.  I guess if this isn't done then a couple of fields won't get initialised.  The way to do that is to change [line 78 of CAnomalyJob.cc](https://github.com/elastic/ml-cpp/blob/d5211a76f9227fb9cbbb44dd10c0126502bc5ef7/lib/api/CAnomalyJob.cc#L78).

droberts195 commented on docs/CHANGELOG.asciidoc:
```diff
@@ -29,6 +29,7 @@
 
 Improve and use periodic boundary condition for seasonal component modeling ({pull}84[#84])
 Improve robustness w.r.t. outliers of detection and initialisation of seasonal components ({pull}90[#90])
+Extend unit testing and make some improvements to deal with changes in seasonal components ({pull}91[#91])
```
I don't think the release notes need to mention unit test changes - someone trying to find out what's new in ML isn't going to care.

tveasey replied:
Fair enough, I'll update.

droberts195 commented on lib/maths/CSeasonalComponent.cc:
```diff
@@ -210,29 +207,18 @@ double CSeasonalComponent::delta(core_t::TTime time,
-        scale = CTools::truncate(1.002 * scale - 0.001, 0.0, 1.0);
-
-        return -scale * min[0] * CTools::sign(shortPeriodValue);
+        return bias.signMargin() != 0.0 ? bias.signMargin() : cancellation.signMargin();
```
Would you want to treat very small numbers as 0?  For example, if `bias.signMargin()` was 1e-308 would you still want to use it?  (This is only a question - if the answer is yes then fine.)

tveasey replied:
One could, for example, not bother if it is << amplitude. The chances of this being very small, but not identically zero, is actually small. If there is no bias then we expect the differences at the different times in the period to *not* all have the same sign with a high probability because noise in the underlying values should cause them to randomly land on different sides of the mean. Also, it doesn't have a significant impact to have a small, but non-zero, delta. So on balance I don't think this warrants the extra complication.

tveasey commented:
This turned out to be more involved: testing with variable decay rate I again hit the instability. I've therefore significantly extended the original PR:
1) I did some more work on the delta calculation. In particular, it minimises the sum component amplitude.
2) I added in active monitoring for signs of instability, by inspecting the sum component amplitude, and reduce the gain if the modelling displays any symptoms of instability.
3) I made sure we remove *all* components if the prediction accuracy drops below the reference.

I split the extra changes into two commits. The first is a pure refactor to better encapsulate seasonal and calendar components. The second makes the functional changes. Can you take another look @droberts195.
